libcontainer/process_linux.go

```diff
@@ -255,6 +255,19 @@ func (p *initProcess) execSetns() error {
 	return nil
 }
 
+// parentError attempts to read any error from the parent so we have better context
+// about the cause of the failure. If no error is found then a new system error for
+// err and cause is returned.
+func parentError(pipe *os.File, err error, cause string) error {
+	if ierr := parseSync(pipe, func(sync *syncT) error { return nil }); ierr != nil {
```

Comment:
> I think there may be a miss-understanding of what parseSync does due to its name. I believe that it is actually just a method to read data from the pipe by aligning on messages that a prefixed by a syncT which act as a message "headers".

`parseSync` parses `syncT`s from a given pipe. Sorry for the confusing name (I wrote it to abstract all of the `syncT` parsing code introduced previously).

> If my understanding is correct parseSync might be better named processMsgs.

That name is more confusing IMO.

> The problem of if the child has died or not is a tricky one. Given the early nature of these failures its highly likely but not guaranteed, so it might be a good idea ensure it is dead by killing it as we're never going to use it again.

Sure, and that's a round-about way of forcing an `EOF`. Still not sure if it makes sense.

> That however is a bigger issue as the whole of this function is fragile with regards to that, as only if it succeeds is the wait and hence full cleanup performed currently. Fixing this would require a much bigger restructure, which I would suggest should be done separately.

I think that's because you're misusing `parseSync` -- its purpose is to be used within the synchronisation state machine code. It's a bit odd how you're using it, and I doubt that it works as intended.